File: typescript/src/code-smells/change-preventers/divergent-change.ts
export type User = { id: string; name: string; email: string }

export class ProfileManager {
  private store = new Map<string, User>()

  register(user: User): void {
    if (!user.name.trim()) throw new Error('invalid name')
    if (!user.email.includes('@')) throw new Error('invalid email')
    this.store.set(user.id, user)
  }

  updateEmail(id: string, newEmail: string): void {
    if (!newEmail.includes('@')) throw new Error('invalid email')
    const u = this.store.get(id)
    if (!u) throw new Error('not found')
    this.store.set(id, {...u, email: newEmail})
  }

  exportAsJson(): string {
    return JSON.stringify(Array.from(this.store.values()))
  }

  exportAsCsv(): string {
    const rows = [
      'id,name,email',
      ...Array.from(this.store.values()).map((u) => `${u.id},${u.name},${u.email}`),
    ]
    return rows.join('\n')
  }

  sendWelcomeEmail(user: User): string {
    return `Welcome ${user.name}! Sent to ${user.email}`
  }
}

export function demoDivergentChange(pm: ProfileManager, u: User): string {
  pm.register(u)
  pm.updateEmail(u.id, u.email)
  return pm.exportAsJson()
}

File: csharp/src/code-smells/change-preventers/DivergentChange.cs
namespace RefactoringAvanzado.CodeSmells.ChangePreventers;

public record User(string Id, string Name, string Email);

public class ProfileManager
{
    private readonly Dictionary<string, User> _store = new();

    public void Register(User user)
    {
        if (string.IsNullOrWhiteSpace(user.Name))
            throw new InvalidOperationException("invalid name");
        if (!user.Email.Contains('@'))
            throw new InvalidOperationException("invalid email");
        _store[user.Id] = user;
    }

    public void UpdateEmail(string id, string newEmail)
    {
        if (!newEmail.Contains('@'))
            throw new InvalidOperationException("invalid email");
        if (!_store.TryGetValue(id, out var u))
            throw new InvalidOperationException("not found");
        _store[id] = u with { Email = newEmail };
    }

    public string ExportAsJson()
    {
        return System.Text.Json.JsonSerializer.Serialize(_store.Values);
    }

    public string ExportAsCsv()
    {
        var rows = new List<string> { "id,name,email" };
        rows.AddRange(_store.Values.Select(u => $"{u.Id},{u.Name},{u.Email}"));
        return string.Join('\n', rows);
    }

    public string SendWelcomeEmail(User user)
    {
        return $"Welcome {user.Name}! Sent to {user.Email}";
    }
}

public static class DivergentChangeDemo
{
    public static string DemoDivergentChange(ProfileManager pm, User u)
    {
        pm.Register(u);
        pm.UpdateEmail(u.Id, u.Email);
        return pm.ExportAsJson();
    }
}
